### PR TITLE
[api] return an error when the specified height is not supported

### LINKED ISSUE #4167: Archive mode related RPC return wrong result
### What version of iotex-core image (or code branch) are you using?
All 

### What did you do? If possible, provide a recipe for reproducing the error.

Query chain' state from block height related RPC, the result always return latest state, not queried height.

```
curl https://babel-api.mainnet.iotex.io \
  -X POST \
  -H "Content-Type: application/json" \
  --data '{"method":"eth_getBalance","params":["0x3c3acce50f7bb6390937139f9e10f99478a83702", "0x1f4"],"id":1,"jsonrpc":"2.0"}'

{"jsonrpc":"2.0","id":1,"result":"0x399a62af592971339dcb"}
```

### What did you expect to see?

```
{"jsonrpc":"2.0","id":1,"error":{"code":-32000,"message":"missing trie node 945e00a1e6f72d0cea125f726a81a110c5f04da829238e186cdcf5e887e8be79 (path )"}}
```

A pruned node cannot serve API requests for certain RPC methods older than 128 blocks.

Should fixed API:

- eth_getBalance
- eth_getCode
- eth_getTransactionCount
- eth_getStorageAt
- eth_call

### PR DESCRIPTION
**Pull request by millken**
# Description
Currently we do not support archive mode. When querying these web3APIs 

- eth_getBalance
- eth_getCode
- eth_getTransactionCount
- eth_getStorageAt
- eth_call

if a height is specified, the result is returned for the latest height. This PR fixes returning an error if height is not supported

Fixes #4167

## Type of change
Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: api/web3server.go [hunk 1]
```diff
@@ -311,11 +312,36 @@ func (svr *web3Handler) getBlockByNumber(in *gjson.Result) (interface{}, error)
 	return svr.getBlockWithTransactions(blk.Block, blk.Receipts, isDetailed.Bool())
 }
 
+func (svr *web3Handler) checkInputBlockNumber(str string) error {
```
Comment (envestcc) [inline api/web3server.go:315]: according to the eth standards, this `str` could be either a block number, a block enum, or a block hash.

Comment (envestcc) [reply api/web3server.go:315]: and prefer name it as `checkInputBlock`

Comment (millken) [reply api/web3server.go:315]: https://ethereum.org/en/developers/docs/apis/json-rpc/#default-block

> When requests are made that act on the state of Ethereum, the last default block parameter determines the height of the block.

> The following options are possible for the defaultBlock parameter:

> HEX String - an integer block number
> String "earliest" for the earliest/genesis block
> String "latest" - for the latest mined block
> String "safe" - for the latest safe head block
> String "finalized" - for the latest finalized block
> String "pending" - for the pending state/transactions

Comment (millken) [reply api/web3server.go:315]: should see this  https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_getbalance

**Comment by ququzone** [comment]
This fix should be released with archive mode features, because this fix will cause graph-node fail without archive mode.

File: api/web3server.go [hunk 2]
```diff
@@ -322,11 +323,46 @@ func (svr *web3Handler) getBlockByNumber(in *gjson.Result) (interface{}, error)
...
 }
 
+func (svr *web3Handler) checkInputBlock(str string) error {
+	switch str {
+	case "", _earliestBlockNumber, _pendingBlockNumber:
+		return errInvalidBlockHeight
+	case _latestBlockNumber:
+		return nil
+	default:
+		//check str is block hash string
```
Comment (CoderZhi) [inline api/web3server.go:333]: `tip := svr.coreService.TipHeight()`

File: api/web3server.go [hunk 3]
```diff
@@ -322,11 +323,46 @@ func (svr *web3Handler) getBlockByNumber(in *gjson.Result) (interface{}, error)
 	return svr.getBlockWithTransactions(blk.Block, blk.Receipts, isDetailed.Bool())
 }
 
+func (svr *web3Handler) checkInputBlock(str string) error {
```
Comment (CoderZhi) [inline api/web3server.go:326]:
```
    func (svr *web3Handler) checkHeight(input gjson.Result) error {
        if !input.Exists() {
             return nil
        }

**Review by CoderZhi — APPROVED** [review]

**Pull request closed without merging**